### ISSUE TITLE
Make Checkstyle multithread save?

### ISSUE THREAD

**Issue by jo2** [checkstyle/checkstyle#7253]
I'd like to synchronize checkstyle code analysis but doing so I run into multiple problems, for example in `` line .. causes a NoSuchElementException, bacause the stack already is empty.

To fix this, checkstyle wouldn't have to implement synchronization, it would just have to be thread save. 

Is something like that planned for one of the next releases or is there something you're working on? If not, I could try fixing the problems I'll run into and create a PR after I'm finished? Would you be interested in that and make it part of one of the next releases (only once it's finished and tested of course)?

**Comment by strkkk** [comment]
@jo2 It is not clear what the problem is. Can you please provide code and config to reproduce your problem?
Btw, it is advised to use issue template to report problems.

**Comment by jo2** [comment]
@strkkk I have multiple calls to a `Checker` object which I'd like to handle using a parallel stream. I'm using checkstyle 8.26. The (first?) problem occurs when I make the `Checker.process()` call.

During the call hierarchy in process the `processIter()` method of the `TreeWalker` is called. `notifyLeave()` calls `leaveToken()` of, in my case, a `ReturnCountCheck` object. The implementation of the `leave()` method in that class is the following:
```
private void leave(DetailAST ast) {
    context.checkCount(ast);
    context = contextStack.pop();
}
```
Because I'm using a parallel stream there are race conditions accessing the stack which causes `contextStack.pop()` to throw a `NoSuchElementException`.

**Comment by romani** [comment]
@jo2 , initially I decided to not participate in this issue ..... but  .... I really appreciate your help with mobile web site (as I use it alot day by day ... to reply all users), so I will try to collect all incomplete conversations between me and @rnveach  about MT mode. 
Problem of MT is rooting too deep to checkstyle architecture and goals that we never had a time to think about it as stream of PRs for functional changes consume almost all time of me and @rnveach .  Users wants functional correctness more than speed, and I completely ok with this.

I need to reply to several PRs first ... as people await for feedback for long time, but  I will be back, please do not hesitate to pin me here if I not reply in 1-2 week.

**Comment by romani** [comment]
Good umbrella for all MT discussions and history is special MT project - https://github.com/checkstyle/checkstyle/projects/5
Proposed implementation of cloning checks instances looks strange to me.

We know that the most time consuming part is parsing of java files.

We have two discussion threads  (read from exact comment till end):
https://github.com/checkstyle/checkstyle/issues/4869#issuecomment-396006851
https://github.com/checkstyle/checkstyle/issues/5140#issuecomment-396005863

Where I and @rnveach can not come to agreement on way to make MT.
@rnveach , looks like was in favor(or ok) with proposed model. 
I was in favor to separate Multifile Checks to separate execution of Checkstyle, and make MT model to parse in one/few threads, process each AST tree in separate instance of Checker(with all Checks). 

As there are more contributors around .... and somebody have time to push this heavy topic forward .... lets continue.

@pbludov  and @strkkk , @jo2 , please share your thoughts on this.

**Comment by pbludov** [comment]
>  I think we can make MT mode for file parsing only (no cloning or ..... , just hidden feature of TreeWalker)

I like that proposal. It can even be divided into three phases: parsing, checking and filtering.

**Comment by romani** [comment]
No, you leaning us to full MT, please read referenced discussion for nuances. 
My latest proposal to keep all as is single thread for all validations/filters and do MT for parsing only ( it will  not be full MT as  our parser is not ready for MT, but we can create separate instance)

**Comment by romani** [comment]
before this issue go back in our backlog .... and I might loose a context again on where we left ....

@rnveach , please let me know if you position is changed through the time.
@strkkk , @jo2 , please share your thoughts on this.

**Comment by rnveach** [comment]
Position hasn't changed. I don't feel MT should be split into 2 runs.

**Comment by romani** [comment]
@rnveach , what about do MT for parsing only ?

**Comment by rnveach** [comment]
It sounds like you want to make TreeWalker have multi-file support which we can't fully endorse at this moment. We still have open issues on that. For parsing alone to be MT, it seems to me the whole thing needs to be MT, or some MT support. I don't see how are you going to send multiple files to be parsed, while keeping Checker single threaded and only working on 1 file at a time and not having anything be returned since the file may or may not be finished.

If it is possible, I am willing to review it.

**Comment by pbludov** [comment]
If I understand Roman correctly, the idea is to make the process two-threaded: the first thread parses the files and puts the ASTs into a queue. The second thread takes the next AST from the queue and processes it.

**Comment by jo2** [comment]
As far as I understood this, the main issue is, that the checks can't be executed in parallel. I couldn't get the reason for this though. According to https://github.com/checkstyle/checkstyle/issues/4869#issuecomment-396006851 there are checks which are executed for each file but after a executing it a check is used up because it has to be cloned for other files?

What matters is the result of the check, right? wouldn't it be possible to have a map containing the filepath as key and a list of checkresults? when checks are started, the value is null on every entry. When check is finished the checkresults are appended to the list of checkresults of a specific file. By doing so every check of every file is independent from each other and they only have to provide the results which are send to a conrolling class. Like:

```
// registered checks containing all classes which implement CheckstyleCheck interface
// can be done with reflection for example
// files are all files whoch should be checked
Map<String, CheckResult> results = new LinkedHashMap();
registeredChecks.parallelStream().forEach(check -> {
  files.parallelStream().forEach(file -> {
    Result result = check.checkFile(rootModule.configure(config), file);
    if (!results.contains(file.getPath()) {
      results.push(file.getPath(), new ArrayList())
    }
    results.get(file.getPath()).add(result);
  });
});    
```
